lib/utils/element.js

```diff
@@ -0,0 +1,193 @@
+var forEach = require('./lodash').forEach
+var isString = require('./lodash').isString
+var disposable = require('./misc').disposable
+
+// Info: http://www.quirksmode.org/js/events_properties.html
+function getEventTarget (event) {
+  // W3C says it's event.target, but IE8 uses event.srcElement
+  var target = event.target || event.srcElement || window.document
+
+  // If an event takes place on an element that contains text, this text node,
+  // and not the element, becomes the target of the event
+  return target.nodeType === 3 ? target.parentNode : target
+}
+
+function addEventListener (el, type, fn) {
+  if (el.addEventListener) {
+    el.addEventListener(type, handler, false)
+    return disposable(function () {
+      el.removeEventListener(type, handler, false)
+    })
+  } else if (el.attachEvent) {
+    el.attachEvent('on' + type, handler)
+    return disposable(function () {
+      el.detachEvent('on' + type, handler)
+    })
+  } else {
+    throw new Error('addEventListener')
+  }
+
+  // IE8 doesn't pass an event param, grab it from the window if it's missing
+  // Calls the real handler with the correct context, even if we don't use it
+  function handler (event) {
+    fn.call(el, event || window.event)
+  }
+}
+
+function shouldIgnoreElement (el) {
+  if (!el || !el.tagName) {
+    return true
+  }
+
+  var tag = el.tagName.toLowerCase()
+  var type = el.getAttribute('type')
+  if (tag === 'input' && type === 'password') {
+    return true
+  }
+
+  var role = el.getAttribute('role')
+  if (
+    role === 'button' ||
+    role === 'link' ||
+    tag === 'a' ||
+    tag === 'button' ||
+    tag === 'input'
+  ) {
+    return false
+  }
+
+  return true
+}
+
+function createTreeHasIgnoreAttribute (ignoreAttribute) {
```

Comment:
Just thinking out loud: Wondering what's the value of `treeHasIgnoreAttribute` being a separate function from `shouldIgnoreElement`? Guess it doesn't matter since we're only calling them internally and so we can ensure that one isn't called without the other- but it seems simpler if `shouldIgnoreElement` also handled ancestry.